### ISSUE TITLE
Restore support for custom values for severity

### ISSUE THREAD

**Issue by nightwing** [eslint/eslint#7394]
Before the introduction of config validator in https://github.com/eslint/eslint/commit/fd8fbe721781ce26957ad0f0b3ffbb33201ac387 it was possible to use any value as severity.
Default eslint formatter treated these values as warning, and json formatter passed value as is allowing tools using eslint to handle some errors in a special way. (e.g. show info markers in the ide)
Would you consider a pr changing [the code](https://github.com/eslint/eslint/blob/660e0918933e6e7fede26bc675a0763a6b357c94/lib/config/config-validator.js#L84-L87) for checking the severity to something like
```js
    validSeverity = (
        (typeof severity === "number") ||
        (typeof severity === "string" && /^(?:off|warn|error|info)$/i.test(severity))
    );
```
---
`What version are you using?` 3.8.1.
`What did you do?` set rule severity to 0.1.
`What happened?` eslint crashed with an error about invalid severity.
`What did you expect to happen?` eslint to work like old versions did.

**Comment by platinumazure** [comment]
What problem are you trying to solve, exactly? It's not clear to me why anyone would need support for this, and (I could be wrong, but) this seems like an XY problem.

**Comment by nightwing** [comment]
@platinumazure when using eslint to dynamically display errors in the editor, it is useful to have info error level for errors that are not important to solve right away.
Previously we could set custom severity value for these rules in eslintrc, and let the editor handle displaying them. Now to achieve same result we have to keep configuration in two separate places.

**Comment by platinumazure** [comment]
@nightwing Okay, so if I'm reading this correctly, the goal is not to add a new level of severity (that ESLint would recognize per se), but rather to support custom severity levels that map to ones ESLint knows about.

I don't like the idea of doing this implicitly (and I *certainly* don't like the idea of doing this by changing a regex to support just one more hardcoded value), but I could possibly support a configuration enhancement to support this. For example:

```json
{
    "customSeverityLevels": {
        "info": "warn"    // Key is the new term, value must be "off"/"warn"/"error"/0/1/2
    },
    "rules": {
        "max-len": ["info", 80],  // Really a "warn"
        "semi-spacing": ["warn", "never"],
        "semi": ["error", "always"]
    }
}
```

That said, the rest of the team may oppose even this, because it's a decent amount of extra code to maintain for (from our perspective) maybe not a lot of value.

@nightwing If you're okay with an explicit, configuration-based proposal for opting into this, then we can evaluate that as a core enhancement. Let me know.

**Comment by nightwing** [comment]
@platinumazure indeed the goal is to restore support for passing arbitrary values for severity from config to the linter output, not change levels supported by eslint itself.
Adding customSeverityLevels attribute seems like a hard sell, because it adds more code and more complexity to the documentation, while making validation less strict would only remove code.

**Comment by platinumazure** [comment]
Well, given that we never intentionally supported custom severity values (and the fact that they worked at all was a complete bug), "only remov[ing] code" is in fact the harder sell. You're asking us not only to restore a bug, but also to do so without adding documentation or otherwise owning that choice.

No, I think if we do anything here, it needs to be explicit and documented. And at this point, I'm not convinced of the value of supporting this in core ESLint.

You can use our CLIEngine API and/or .js configuration to implement this yourself. You could have a raw configuration file with all the configuration values you want, and then either a .js configuration file or your CLIEngine-consuming code would transform the configuration to match ESLint's supported levels. Then, after linting and fixing, your CLIEngine-consuming code would iterate through all messages, look up the original configuration by rule ID, and transform the severity accordingly.

**Comment by nightwing** [comment]
Unfortunately the approach with CLIEngine API is not useful here since it still means, users need to have two configuration files, one for the ide and one for the eslint.

**Comment by platinumazure** [comment]
@nightwing The approach I had proposed earlier (adding a `customSeverityLevels` section to config files) only requires one configuration file. Feel free to reopen if you would like to continue discussing that approach, otherwise thanks for the discussion.
